Refactor and extend quarantine implementation in twister

gchwier opened the pull request:
Implementation ported from TwisterV2.
- quarantine handled by separate module
- multiple yaml allowed from args: --quarantine-list
  example: `scripts/twister -T tests/kernel/common -p native_posix --quarantine-list test_quarantine1.yml --quarantine-list test_quarantine2.yml`
- scenarios, platforms, architectures keywords in quarantine yaml are optional, if not given - means take them all.
  Means, to filter all scenarios for platforms `qemu_cortex_m3` and `native_posix`, one can use
```
- platforms:
  - qemu_cortex_m3
  - native_posix
 comment: "reason"
```
  or
```
- scenarios:
  - all
 platforms:
  - qemu_cortex_m3
  - native_posix
 comment: "reason"
```
- `architecture` keyword also added like in solution proposed in PR #52179

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier committed:
twister: Refactor and extend quarantine implementation in twister

Implementation ported from TwisterV2.
- quarantine handled by separate module
- multiple yaml allowed from args: --quarantine-list
- scenarios, platforms, architectures keywords in
quarantine yaml are optional, if not given - means take it all

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

nashif commented:
> - scenarios:
>   - all
>  platforms:
>   - x86
>   - arm
>  comment: "reason"

thanks, I am ok with the new implementation and separate "module" for dealing with quarantine, but IMO we are overloading and misusing the current syntax and scheme. We wanted to add more "fields" (simulators, toolchains, etc.) where we want to apply quarantine and having all this fit under the existing scheme seems excessive and will not be easy to understand or read.

Either way, to replace the existing PR, this will need to support the other features added in #52179:
- simulators
- regex support on test names

I wanted to add toolchains as well in a future PR.

gchwier committed:
twister: quarantine: Added simulations field and regex to filter fields

Regex are allowed on every field, not only scenarios.
Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier commented:
I added `simulations` field. Regex is also used (for every quarantine filter fields: scenarios, platforms, architectures and simulations).
Example quarantine file:
```
- scenarios:
  - kernel.common.(misra|tls)
  comment: "link to issue"

# quarantine kernel.common scenario for all platforms with name started from nrf53
- scenarios:
  - kernel.common
  platforms:
  - nrf53.*

# quarantine all scenarios where simulation is qemu, without reason and no regex
- simulations:
  - qemu
```

PerMac commented:
@nashif  If I understand correctly, your are fine with the proposed refactoring, as long as all the required features will be added? Grzegorz added features you suggested. 

I guess what also we need to discuss is ` IMO we are overloading and misusing the current syntax and scheme` i.e. How the scheme should look like. I went through your PR #52179 and I think I see what you meant. You are proposing some additional level of organization, where platforms/simulation/toolchains etc. will have extra dedicated entries. At first I found it useful (but problematic since it will break our CIs and some simultaneous merges will have to follow to adapt CI). However, after longer consideration I think the solution Grzegorz proposes will serve us better.

In Grzegorz's solution an entry in quarantine yaml is more flexible, and I will argue why it is good for us: All used keys (scenario/platform/simulation/etc) can be mixed together creating a desired combination of what to quarantine.  Also, regexes can be used to all fields, not only scenarios. If in a given entry a certain key (e.g. scenario) is missing it means no specific value chosen, i.e. take all possible.  So having everything skipped on a given platform or simulation is also possible. This was pointed by Grzegorz in his latest comment.

I think the schema you're proposing might quickly end as obsolete and would require yet another modification. E.g. What if you'd like to combine scenarios with other fields, e.g. simulation? I can see a case when we don't want to skip all `simulation: qemu` but e.g. only logging related tests on qemu type platforms. Or (in future) all drivers tests on a given toolchain. In your schema all new keys have their own sections and cannot be used in combination with any other key.

As for organization and not introducing a mess in the quarantine file. I think, if needed, we can use comments to organize the file. They will be skipped by twister, but can help humans to organize entries. E.g. 

```
###### Mixed keys  ##################################
- scenarios:
  - kernel.common.(misra|tls)
  comment: "link to issue"

- scenarios:
  - kernel.common
  platforms:
  - nrf53.*

####### Platforms  ##################################
### Use this section to store platforms to be fully skipped ######

- platforms:
  - nrf99.*
  comment: All nrf99 family boards got their dts corrupted. Issue: xxxx
   

####### Simulators ##################################
### Use this section to store simulators to be fully skipped ######
- simulations:
  - qemu
 
```

Also, by definition quarantine should be used as a temporary solution, so the yaml will be a living document, where after fixing entries are dropped. So it won't keep growing indefinitely.
@nashif what do you think? Do my arguments make sense?

nashif reviewed (changes requested):
I am fine with this, however, I would like to see some tests added now that the structure is more complex. Please provide some tests.

nashif commented:
also, please update documentation.

gchwier committed:
twister: quarantine: Added pytests

Implemented scenarios to test the quarantine in twister
Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier committed:
twister: docs: Update twister docs with quarantine

Docs updated with changes proposed in #52179
Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

gchwier commented:
@PerMac @nashif Thank you for your comments.
I updated tests (couple of scenarios with examples of quarantine files).
Updated also documentation with changes proposed in #52179

PerMac reviewed (changes requested):
I found 2 issues:. Sorry that I'm reporting it just now, somehow I missed it earlier on...

I found a small difference in the reporting comparing with the existing one. In this PR one gets:
`"reason":"MY COMMENT",` if `comment` was provided and 
`"reason":"under quarantine",` without comment.

The existing one will report:
`"reason":"Quarantine: MY COMMENT",`  and
`"reason":"Quarantine: NA"` respectively. It is important to have `"Quarantine:" always in the reason because we don't have a dedicated status for such skips. But they can be reliably identify by presence of 
```
           "status":"filtered",
            "reason":"Quarantine: ...",
```
E.g. internally we have a parser analyzing twister reports and looking for the above combination. Matched entries are placed in a separate list highlighting only quarantines. Without "Quarantine" word in the reason there is no reliable way to differentiate what was quarantined from other filtered entries and we need to treat quarantine in a special way. 

This is very simple change. The other might require more thinking:
Quarantine and integration mode:
When we use integration mode (--integration/-G) the scope is chosen from integration_platforms from yaml. It also modifies the workflow: skips on integration_platforms are translated to errors. However, quarantined tests got exemption from this. Our (both upstreams and internal) CIs uses this mode for on-commit runs. This PR removes such exemption and quarantined integration platforms get errors instead of skips in the integration mode. This has to be aligned with the current workflow to make quarantine usable in the CIs.

PerMac commented:
I think adding "Quarantine" to the reason might solve both issues, as the exemption is based on the reason:
https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib/testplan.py#L830

gchwier commented:
@PerMac thank you for finding this issue. I fixed it. Please check it once again.

PerMac reviewed (approved):
Thanks, comments addressed.
